---
title: "cmvdrg-intro"
author: "OJ Charles"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{cmvdrg-intro}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


```{r setup}
library(cmvdrg)
```


## Call resistance mutations from cmvdrg
```{r}
# load some test data, bundled in the package
my_sample = system.file("testdata", "A10.vcf", package = "cmvdrg")

# this calls any variants in your sample, that are also present in the cmvdrg database.
data = call_resistance(infile = my_sample, all_mutations = F)

# view the first few lines, there are a lot of columns!
head(data)

```


## View all mutations

It may be the case that your sequences contain non synonymous mutations in resistance genes, that may confer changes in resultant protein. Although there are a large number of mutations accounted for in this databse, ofcourse there is a possiblity to observe something novel. The call_resistance function allows for these sorts of questions.
```{r}

# returns all variants, with any resistance data, with annotation of protein coded for & if it coded for a change in protein sequence.
mutations_all = call_resistance(infile = my_sample, all_mutations = T)


# to view all mutations in resistance genes we can use base R to filter.
mutations_res = mutations_all[mutations_all$GENEID %in% c("UL54", "UL97", "UL27", "UL51", "UL56", "UL89"),]


# are there any non-synonymous (DNA variants that result in a change of amino acid) variants in resistance genes.
mutations_res_nonsyn = mutations_res[mutations_res$CONSEQUENCE == "nonsynonymous",]

```

as you can see, there are a number of variants in resistance genes that result in a change of protein sequence, and may deserve attention.



## Run the shiny app
```{R}
# runShinyCMV()

```